xchange-coinbasepro/src/main/java/info/bitrich/xchangestream/coinbasepro/CoinbaseProStreamingMarketDataService.java

```diff
@@ -63,7 +63,8 @@ private boolean containsPair(List<CurrencyPair> pairs, CurrencyPair pair) {
         return getRawWebSocketTransactions(currencyPair, false)
                 .filter(message -> message.getType().equals(SNAPSHOT) || message.getType().equals(L2UPDATE))
                 .map(s -> {
-                    if (s.getType().equals(SNAPSHOT)) {
+                    // check if bids/asks contains ccy - in-case we did not get the snapshot (this happens during reconnects)
+                    if (s.getType().equals(SNAPSHOT) || !bids.containsKey(currencyPair) || !asks.containsKey(currencyPair)) {
```

Comment:
Isn't this just going to result in us writing a partial orderbook for the pair?  I wonder if it might be better just to _ignore_ (filter out) updates like these until we get a valid snapshot.

Reply:
yes, i will produce a partial book until SNAPSHOT arrives.  I am good with ignoring too, my goal was to avoid the NPE that occured later down the line, when only one side of the book was present.   Also, I need to take another run thru this code as I am getting some strange results with crossed books, where the L2UPDATE is the only content of the book.  it may be the SNAPSHOT never arrives.

Reply:
OK.  Yes - I think I prefer ignoring and not starting to send orderbooks to the client until we have a clean snapshot, so if you could change that, great.

Good luck finding the other issue.

Reply:
As I stand today, we dont get a clean SNAPSHOT, it gets losts and L2UPDATES pour in.  Is it possible the caching mech is allowing that?, it only seems to hold one last message.
@badgerwithagun

Reply:
closing